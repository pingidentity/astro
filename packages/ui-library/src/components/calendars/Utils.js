var Const = require("./Constants");
var moment = require("moment-range");
var KeyboardUtils = require("../../util/KeyboardUtils.js");

var _keyDownViewHelper = [
    {
        prev: false,
        next: true,
        exit: true,
        unit: "day",
        upDown: 7
    },
    {
        prev: true,
        next: true,
        unit: "months",
        upDown: 3
    },
    {
        prev: true,
        next: false,
        unit: "years",
        upDown: 3
    }
];

module.exports = {

    keyDownActions: function (code) {
        var _viewHelper = _keyDownViewHelper[this.state.currentView];
        var unit = _viewHelper.unit;
        // Use today's date if no date given
        var date = this.state.date || moment();
        var newDate;

        switch (code) {
            case Const.keys.left:
                newDate = date.clone().subtract(1, unit);
                break;
            case Const.keys.right:
                newDate = date.clone().add(1, unit);
                break;
            case Const.keys.up:
                newDate = date.clone().subtract(_viewHelper.upDown, unit);
                break;
            case Const.keys.down:
                newDate = date.clone().add(_viewHelper.upDown, unit);
                break;
            case Const.keys.enter:
                if (_viewHelper.prev) {
                    switch (unit) {
                        case "months":
                            newDate = module.exports.getNearestInRange(date, this.props.dateRange);
                            break;
                        case "years":
                            newDate = module.exports.getNearestInRange(date, this.props.dateRange, "months");
                            break;
                    }
                    this.prevView(newDate || date);
                }

                if (_viewHelper.exit) {
                    this.setState( { isVisible: false } );
                }

                break;
            case Const.keys.esc:
                this.setState( { isVisible: false } );
                break;
        }
        if (KeyboardUtils.isArrowKey(code) && module.exports.inDateRange(newDate, this.props.dateRange, unit)) {
            this.setDate(newDate);
        }
    },

    inDateRange: function (date, dateRange, unit) {
        if (!date || !dateRange) {
            return true;
        }
        var current = moment(date),
            start = dateRange.startDate && moment(dateRange.startDate),
            end = dateRange.endDate && moment(dateRange.endDate);

        switch (unit) {
            case "months":
                // Disregard different date of month when comparing months by setting all to 1
                if ((start && current.clone().date(1).isBefore(start.date(1))) ||
                        (end && current.clone().date(1).isAfter(end.date(1)))) {
                    return false;
                }
                break;
            case "years":
                if ((start && current.clone().year() < start.year()) || (end && current.clone().year() > end.year())) {
                    return false;
                }
                break;
            default:
                if ((start && current.isBefore(start)) || (end && current.isAfter(end))) {
                    return false;
                }
                break;
        }
        return true;
    },

    getNearestInRange: function (date, dateRange, unit) {
        if (module.exports.inDateRange(date, dateRange, unit)) {
            return date;
        } else {
            var start = dateRange && dateRange.startDate && moment(dateRange.startDate),
                end = dateRange && dateRange.endDate && moment(dateRange.endDate),
                nearest = date.clone();

            switch (unit) {
                case "months":
                    // Set to nearest month within in the start/end range
                    if (start && (start.year() === date.year())) {
                        nearest = date.month(start.month());
                    } else if (end && (end.year() === date.year())) {
                        nearest = date.month(end.month());
                    }
                    break;
                case "years":
                    // Set to nearest year within in the start/end range
                    if (start && (date.year() < start.year())) {
                        nearest = date.year(start.year());
                    } else if (end && (date.year() > end.year())) {
                        nearest = date.year(end.year());
                    }
                    break;
                default:
                    // Set to nearest month within in the start/end range
                    if (start && (start.month() === date.month())) {
                        nearest = date.date(start.date());
                    } else if (end && (end.month() === date.month())) {
                        nearest = date.date(end.date());
                    }
                    break;
            }
            return nearest;
        }
    }

};