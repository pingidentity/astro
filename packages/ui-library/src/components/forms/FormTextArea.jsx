"use strict";

var React = require("react"),
    css = require("classnames"),
    FormFieldConstants = require("../../constants/FormFieldConstants"),
    HelpHint = require("../tooltips/HelpHint.jsx"),
    _ = require("underscore");

/**
 * @module FormTextArea
 * @desc A text area component that supports edit and readonly modes.
 *
 * @param {string} [referenceName] name used by the parent to get the value/text entered in this component
 *      (default 'form-text-area')
 * @param {boolean} [isRequired] true if the field is required (default false)
 * @param {string} [value] current text field value, used when state is managed outside of component;
 *      must be used with onValueChange handler to get updates
 * @param {string} [defaultValue] the default (initial) value to be shown in the field, when component managing state
 *      itself
 * @param {string} [labelText] the text to show as the field's label
 * @param {string} [placeholder] input placeholder text
 * @param {function} [onValueChange] a callback that will be triggered when the field changes
 * @param {function} [onBlur] a callback that will be triggered when the field loses focus
 * @param {string} [mode] how the field will be shown: edit or readonly (default edit)
 * @param {number} [maxLength] max length (number) supported by the field
 * @param {number} [cols] columns value for sizing, default is css
 * @param {number} [rows] rows value for sizing, default is css
 * @param {string} [originalValue] value to set the field to if the undo icon is clicked.
 *      (if not specified, no undo control will be shown)
 * @param {boolean} [useAutocomplete] whether or not the field will support autocomplete. (default false)
 * @param {string} [inputCss] CSS classes to add to the input element
 * @param {string} [className] CSS classes to add to the parent Label element
 * @param {string} [errorMessage] an error message to show (will be shown if defined)
 *
 * @example
 *                   <FormTextArea
 *                       referenceName={name}
 *                       labelText={element.viewName}
 *                       isRequired={element.required}
 *                       defaultValue={element ? element.value : ''}
 *                       onValueChange={myFunction} />
 */
var FormTextArea = React.createClass({

    propTypes: {
        // prop validations
        mode: React.PropTypes.string,
        isRequired: React.PropTypes.bool,
        defaultValue: React.PropTypes.string,
        value: React.PropTypes.string,
        labelText: React.PropTypes.string,
        labelHelpText: React.PropTypes.string,
        placeholder: React.PropTypes.string,
        onValueChange: React.PropTypes.func,
        onBlur: React.PropTypes.func,
        referenceName: React.PropTypes.string,
        maxLength: React.PropTypes.number,
        originalValue: React.PropTypes.string,
        cols: React.PropTypes.number,
        rows: React.PropTypes.number,
        useAutocomplete: React.PropTypes.bool,
        inputCss: React.PropTypes.string,
        className: React.PropTypes.string,
        errorMessage: React.PropTypes.string
    },

    /**
     * Handle a change of the field's value.
     *
     * @param {object} e the event object
     * @private
     */
    _handleFieldChange: function (e) {
        // return the event to the parent so it can perform whatever operations it needs
        this.props.onValueChange(e);
        // set the internal state to reflect the new value
        this.setState({
            fieldValue: e.target.value
        });
    },

    /**
     * Handle the field losing focus.
     *
     * @param {object} e the event object
     * @private
     */
    _handleBlur: function (e) {
        this.props.onBlur(e);
    },

    /**
     * Handle the undo operation by changing the value back to originalValue.
     *
     * @param {object} e the event object
     * @private
     */
    _handleUndo: function (e) {
        var originalValue = this.props.originalValue;
        // update the event with the reverted data and send back to the parent
        // otherwise the parent won't be aware of the reverted field even though the browser displays the original value
        e.target = this.refs[this.props.referenceName].getDOMNode();
        e.target.value = originalValue;
        this.props.onValueChange(e);
        // trigger a re-rendering of the field so the 'undo' link dissapears and the value gets reverted
        this.setState({
            fieldValue: originalValue
        });
    },

    getDefaultProps: function () {
        return {
            referenceName: "form-text-area",
            className: "input-textarea",
            mode: FormFieldConstants.FormFieldMode.EDIT,
            defaultValue: "",
            labelText: "",
            useAutocomplete: false,
            onValueChange: function () {
                // do nothing as the default action
            },
            onBlur: function () {
                // do nothing as the default action
            }
        };
    },

    getInitialState: function () {
        return {
            fieldValue: this.props.defaultValue
        };
    },

    render: function () {
        var value = _.isUndefined(this.props.value) ? this.state.fieldValue : this.props.value,
            edited = this.props.originalValue && this.props.originalValue !== value,
            readonly = this.props.mode.toUpperCase() === FormFieldConstants.FormFieldMode.READ_ONLY,
            labelHelp;

        var labelCss = css(this.props.className, {
                "input-textarea": true,
                required: this.props.isRequired,
                "form-error": this.props.errorMessage,
                edited: edited,
                "value-entered": !!value,
                readonly: readonly
            }),
            errorCss = css("help-tooltip form-error-message", {
                show: this.props.errorMessage
            });

        var undo;
        if (this.props.originalValue !== null &&
                this.props.originalValue !== undefined &&
                (value !== this.props.originalValue)) {
            // only show the undo icon if an original value is passed in and the text area's value has changed
            // empty strings are OK
            undo = (<a data-id="undo" className="undo" onClick={this._handleUndo}>undo</a>);
        }

        if (this.props.labelHelpText) {
            labelHelp = (
                <HelpHint hintText={this.props.labelHelpText} data-id={this.props.referenceName + "_helptooltip"} />
            );
        }

        return (
            <label className={labelCss}>
                <span className="label-text">
                    {this.props.labelText}
                    {labelHelp}
                </span>
                <span className="input-container">
                    <textarea
                        className={this.props.inputCss}
                        placeholder={this.props.placeholder}
                        isRequired={this.props.isRequired}
                        ref={this.props.referenceName}
                        data-id={this.props.referenceName}
                        name={this.props.referenceName}
                        readOnly={readonly}
                        maxLength={this.props.maxLength}
                        value={value}
                        cols={this.props.cols}
                        rows={this.props.rows}
                        autoComplete={this.props.useAutocomplete ? "on" : "off"}
                        onChange={this._handleFieldChange}
                        onBlur={this._handleBlur}>
                    </textarea>
                    {undo}
                    <div className={errorCss}>
                        <div className="tooltip-text" data-id={this.props.referenceName + "_errormessage"}>
                            {this.props.errorMessage}
                        </div>
                    </div>
                </span>
                {this.props.children}
            </label>
        );
    }

});

module.exports = FormTextArea;
