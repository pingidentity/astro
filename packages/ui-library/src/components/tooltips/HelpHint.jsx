import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import _ from "underscore";
import Tippy from "@tippyjs/react";
import Link from "../general/Link";

/**
 * @enum {string}
 * @alias HelpHint.Placements
 */
const Placements = {
    /** top */
    TOP: "top",
    /** top-start */
    TOP_START: "top-start",
    /** top-end */
    TOP_END: "top-end",
    /** bottom */
    BOTTOM: "bottom",
    /** bottom-start */
    BOTTOM_START: "bottom-start",
    /** bottom-end */
    BOTTOM_END: "bottom-end",
    /** left */
    LEFT: "left",
    /** left-start */
    LEFT_START: "left-start",
    /** left-end */
    LEFT_END: "left-end",
    /** right */
    RIGHT: "right",
    /** right-start */
    RIGHT_START: "right-start",
    /** right-end */
    RIGHT_END: "right-end",
};

/**
 * @enum {string}
 * @alias HelpHint.Types
 */
const Types = {
    DEFAULT: "",
    /** error */
    ERROR: "error",
    /** light */
    LIGHT: "light",
    /** success */
    SUCCESS: "success",
    /** warning */
    WARNING: "warning",
};


/**
 * @class HelpHint
 * @desc HelpHint can appear above, to the right, to the bottom or to the left of any DOM element.
 *     By default the hint appears to the upper right of the icon or target
 *     The following css classes may be added to effect the tooltip placement and state:
 *         'right': tooltip displays to the left of the icon instead of the right
 *         'bottom': tooltip displays under the help icon instead of above
 *         'show': keeps tooltip visible regardless of hover
 *     The following css classes maybe added to change the appearance of the tooltip:
 *         'error'/'warning': tooltip shows in red
 *         'success': tooltip shows in green
 *     Multiple rows of text are supported.
 *     HTML is supported.
 *
 * @param {string} [className]
 *     CSS classes to set on the top-level HTML container.
 * @param {string} [data-id="helpHint"]
 *     To define the base "data-id" value for top-level HTML container.
 * @param {number} [delayHide=400]
 *     Sets the number of milliseconds to wait before hiding the tooltip.
 * @param {number} [delayShow=0]
 *     Sets the number of milliseconds to wait before showing the tooltip.
 *
 * @param {any} hintText
 *     Provides the text that will appear in the hint.
 *
 * @param {string} link
 *     Provides a URL for a "More on this topic" link at the bottom of the tooltip.
 * @param {Object} linkProps
 *     Props object that is passed into the Link node when link is present.
 * @param {HelpHint.Placements} [placement]
 *     How to place the help hint.
 * @param {HelpHint.Types} [type]
 *     Sets special styling for the help hint.
 * @param {bool} [unstyleTrigger=false]
 *     When set, don't apply any styling to the helphint trigger.
 *
 *  @example
 *     <HelpHint className="short-tooltip right" hintText="My first HelpHint!">SomeTextWithHelp</HelpHint>
 */

class HelpHint extends React.Component {

    static propTypes = {
        className: PropTypes.string,
        containerClassName: PropTypes.string,
        "data-id": PropTypes.string,
        delayHide: PropTypes.number,
        delayShow: PropTypes.number,
        hintText: PropTypes.any.isRequired,
        iconName: PropTypes.string,
        leftMargin: PropTypes.bool,
        link: PropTypes.string,
        linkProps: PropTypes.shape({}),
        placement: PropTypes.oneOf(Object.values(Placements)),
        type: PropTypes.oneOf(Object.values(Types)),
        tooltipProps: PropTypes.object,
        unstyleTrigger: PropTypes.bool,
    };

    static defaultProps = {
        "data-id": "helpHint",
        delayHide: 400,
        delayShow: 0,
        className: "",
        leftMargin: false,
        unstyleTrigger: false,
        placement: "right",
    };

    _handleClick = (e) => {
        // Kill click event to prevent event from triggering label from checking a checkbox/radio
        e.preventDefault();
    };

    _getPlacement = () => {
        const { classNames, placement: propsPlacement } = this.props;
        let placement = "right";

        // use newer placement prop if present
        if (propsPlacement) {
            placement = propsPlacement;

        // otherwise try parsing it from the className prop
        // since react tooltip only has 4 postions, precedence is given to top/bottom over left/right
        } else if (classNames) {
            placement = classNames.indexOf("left") > -1 ? "left" : placement;
            placement = classNames.indexOf("top") > -1 ? "top" : placement;
            placement = classNames.indexOf("bottom") > -1 ? "bottom" : placement;
        }

        return placement;
    };

    _getTypeClass = () => {
        return this.props.type ? `help-tooltip--${this.props.type}` : null;
    }

    componentDidMount() {
        if (this.props.show || this.props.className.indexOf("show") > -1) {
            this.show();
        }
    }

    getIconName() {
        const {
            iconName,
            lock
        } = this.props;

        if (iconName) {
            return `icon-${iconName}`;
        } else {
            return lock
                ? "icon-lock"
                : "icon-help";
        }
    }

    maybeRenderLink() {
        const { link, linkProps } = this.props;
        return (link || linkProps) &&
            <Link title="More on this topic" url={link} icon="info" type="block" {...linkProps} />;
    }

    maybeRenderExtraContainer = children => (
        this.props.unstyleTrigger
            ? <div className="help-tooltip help-tooltip--extra">{children}</div>
            : children
    );

    tooltipClassName = this.props.link ? classnames("tooltip-text", "tooltip-text-link") : "tooltip-text";

    render() {
        const {
            "data-id": dataId,
            children,
            containerClassName,
            leftMargin,
            unstyleTrigger,
        } = this.props;

        const iconName = this.getIconName(),
            uid = _.uniqueId("rtt_"),
            display = children
                ? children
                : <span className={iconName} data-id={dataId + "-icon"} />;

        const containerClassNames = classnames("help-tooltip", {
            [containerClassName]: containerClassName !== undefined,
            "help-tooltip--left-margin": leftMargin,
        });

        const tippyClassNames = classnames(this._getTypeClass(), {
            "help-tooltip": !unstyleTrigger,

        });

        return (
            <div className={containerClassNames} data-id={dataId}>
                <Tippy
                    placement={this.props.placement}
                    className={tippyClassNames}
                    interactive
                    content={
                        <span
                            id={uid}
                            className={this.tooltipClassName}
                        >
                            {this.props.hintText}
                            {this.maybeRenderLink()}
                        </span>
                    }
                    appendTo={document.body}
                    theme={this.props.type}
                    delay={[this.props.delayShow, this.props.delayHide]}
                    {...this.props.tooltipProps}
                >
                    <div
                        data-id={dataId + "-target"}
                        className={classnames(this.props.className, "help-tooltip-target", {
                            "inline": (!children && !this.props.className), // make the icon inline if no other class is specified
                        })}
                        onClick={this._handleClick}
                        data-tip={true}
                        data-for={uid}
                        ref={(target) => { this.target = target; }}
                    >
                        {display}
                    </div>
                </Tippy>
            </div>
        );
    }
}

module.exports = HelpHint;
module.exports.Placements = Placements;
module.exports.Types = Types;
module.exports.placements = Placements;
module.exports.types = Types;
