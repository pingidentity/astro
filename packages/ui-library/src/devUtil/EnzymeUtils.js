import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
