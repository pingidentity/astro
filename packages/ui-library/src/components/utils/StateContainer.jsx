import React from "react";
import PropTypes from "prop-types";
import Utils from "../../util/Utils";

const mapNothing = () => ({});

const defaultGetPassedProps = props => props;
const containerGetPassedProps = props => props.passedProps;

export const createProgressiveState = stateDefs => ({
    passedProps,
    initialState,
    boundSetState,
    getPassedProps = defaultGetPassedProps,
}) => {
    const isControlled = name => passedProps[name] === undefined;

    // figure out which props will be controlled and create initial state for those
    const pState = stateDefs.reduce((state, { name, initial }) => {
        if (isControlled(name)) {
            return {
                ...state,
                [name]: initialState[name] === undefined ? initial : initialState[name],
                // the initialState prop should override the defined initial value
                // but only if this is a controlled prop
            };
        }
        return state;
    }, {});

    // make sure the keys provided in initialState are actually being controlled
    if (!Utils.isProduction()) {
        Object.keys(initialState).forEach(key => {
            if (pState[key] === undefined) {
                console.warn(
                    `Initial state provided for prop "${key}" that is not being controlled by the component`
                );
            }
        });
    }

    // generate custom callbacks with transform functions
    const makeTransformCallbacks = (name, callbacks = []) =>
        callbacks.reduce(
            (transformCallbacks, { name: cbName, transform, passTransformedValue = false, }) => ({
                ...transformCallbacks,
                [cbName]: (value, e) => {
                    // set the state with the transformed value
                    boundSetState((state, props) => {
                        const transformedValue = transform(value, state[name], getPassedProps(props));

                        if (passedProps[cbName]) {
                            passedProps[cbName](passTransformedValue ? transformedValue : value, e);
                        }

                        return { [name]: transformedValue };
                    });
                }
            }), {}
        );

    // get the callbacks for a single state definition
    const makeCallbacksFromDef = ({ name, setter, callbacks }) => {
        // add any custom callbacks
        const customTransformCallbacks = makeTransformCallbacks(name, callbacks);

        // return with setter if defined
        return setter ? {
            [setter]: (value, e) => {
                boundSetState({ [name]: value });
                // if a callback is provided, we'll still execute it
                if (passedProps[setter]) {
                    passedProps[setter](value, e);
                }
            },
            ...customTransformCallbacks
        } : customTransformCallbacks;
    };

    // create all the callbacks
    const pCallbacks = stateDefs.reduce((returnCallbacks, def) => {
        if (!isControlled(def.name)) {
            return returnCallbacks;
        }
        return { ...returnCallbacks, ...makeCallbacksFromDef(def) };
    }, {});

    return [pState, pCallbacks];
};

class StateContainer extends React.Component {
    static propTypes = {
        stateDefs: PropTypes.arrayOf(
            PropTypes.shape({
                name: PropTypes.string.isRequired, // name of the props
                initial: PropTypes.any.isRequired, // starting value
                setter: PropTypes.string, // name of the callback in the target component
                callbacks: PropTypes.arrayOf( // when you need something different than a simple setter
                    PropTypes.shape({
                        name: PropTypes.string,
                        // Decides whether to pass the transformed value to the callback or just the value that
                        // the base component passes.
                        passTransformedValue: PropTypes.bool,
                        transform: PropTypes.func, // this function will be applied to the value and current state
                    })
                ),
            })
        ),
        initialState: PropTypes.object,
        passedProps: PropTypes.object,
        mapToProps: PropTypes.func,
    };

    static defaultProps = {
        stateDefs: [],
        initialState: {},
        passedProps: {},
        mapToProps: mapNothing,
    };

    constructor({ stateDefs, initialState, passedProps }) {
        // the constructor has two main goals
        // 1) populate the state for any props that will be controlled
        // 2) create callback functions for those props
        super();

        const [ state, callbacks ] = createProgressiveState(stateDefs)({
            initialState, passedProps, boundSetState: this.setState.bind(this), getPassedProps: containerGetPassedProps
        });

        this.state = state;
        this.callbacks = callbacks;
    }

    render() {
        const { children, mapToProps, passedProps } = this.props;

        // alert dev if there are state/prop conflicts
        if (!Utils.isProduction()) {
            Object.keys(this.state).forEach(key => {
                if (passedProps[key] !== undefined) {
                    console.warn(
                        `The prop "${key}" is being controlled by the component because it was not initially set. ` +
                        `The newly-received value will be ignored.`
                    );
                }
            });
        }

        return children({ ...passedProps, ...this.callbacks, ...this.state, ...mapToProps(this.state, passedProps) });
    }
}

export const inStateContainer = (stateDefs, mapToProps) => WrappedComponent => ({ initialState, ...props }) => (
    <StateContainer stateDefs={stateDefs} mapToProps={mapToProps} initialState={initialState} passedProps={props}>
        {containerProps => <WrappedComponent {...containerProps} />}
    </StateContainer>
);

export const toggleTransform = (value, current) => (!current);

export default StateContainer;