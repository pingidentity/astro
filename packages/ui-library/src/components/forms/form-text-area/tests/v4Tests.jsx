import React from "react";
import { mountSnapshotDataIds } from "../../../../devUtil/EnzymeUtils";
import ReactTestUtils from "react-dom/test-utils";
import TestUtils from "../../../../testutil/TestUtils";
import FormTextArea from "../v2";
import HelpHint from "../../../tooltips/HelpHint";
import Utils from "../../../../util/Utils";
import _ from "underscore";
import StateContainer from "../../../utils/StateContainer";
import { mount } from "enzyme";

describe("FormTextArea v4", function () {

    function getComponent (props) {
        props = _.defaults(props || {}, {
            stateless: true,
            onChange: jest.fn(),
            onValueChange: jest.fn(),
            onBlur: jest.fn(),
            flags: [ "p-stateful" ],
        });

        return ReactTestUtils.renderIntoDocument(<FormTextArea {...props} />);
    }

    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <FormTextArea
                stateless={true}
                onChange={jest.fn()}
                onValueChange={jest.fn()}
                onBlur={jest.fn()}
                flags={[ "p-stateful" ]}
            />
        );
    });

    it("renders the component", function () {
        var component = getComponent();

        // verify that the component is rendered
        var field = TestUtils.findRenderedDOMNodeWithClass(component, "input-textarea");
        expect(ReactTestUtils.isDOMComponent(field)).toBeTruthy();
        // make sure that the field is not required by default
        var elements = TestUtils.scryRenderedDOMNodesWithClass(component, "required");
        expect(elements.length).toBe(0);
    });

    it("renders with given value", function () {
        var component = getComponent({
            value: "myValue"
        });

        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        expect(field.value).toBe("myValue");
    });

    it("shows the default value", function () {
        var defaultValue = "my random value";
        var component = getComponent({ value: defaultValue });

        // verify that the component is rendered
        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        expect(field.value).toContain(defaultValue);
    });

    it("shows placeholder", function () {
        var defaultValue = "my random value",
            placeholder = "edit me";

        var component = getComponent({
            defaultValue: defaultValue,
            placeholder: placeholder
        });

        // verify that the component is rendered
        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        expect(field.getAttribute("placeholder")).toEqual(placeholder);
    });

    it("shows labelHelpText", function () {
        var helpText = "help!";
        var component = getComponent({
            labelText: "label",
            labelHelpText: helpText
        });

        var help = TestUtils.findRenderedComponentWithType(component, HelpHint);

        expect(help).toBeTruthy();
    });

    it("shows labelText", function () {
        var component = getComponent({ labelText: "myLabel" });

        var label = TestUtils.findRenderedDOMNodeWithClass(component, "label-text");

        expect(label).toBeDefined();
    });

    it("respects value over defaultValue and state precedence", function () {
        var component = getComponent({
            defaultValue: "my random value",
            value: "my value"
        });

        // verify that the component is rendered
        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        expect(field.value).toContain("my value");

        ReactTestUtils.Simulate.change(field, { target: { value: "abc" } });
        expect(field.value).toContain("my value");
    });

    it("fires the onChange and onValueChange callbacks when field changes", function () {
        var component = getComponent();

        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        ReactTestUtils.Simulate.change(field, { target: { value: "abc" } } );
        expect(component.props.onChange).toBeCalled();
        expect(component.props.onValueChange).toBeCalled();
    });

    it("triggers the onBlur callback on field blur", function () {
        var component = getComponent();

        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");

        ReactTestUtils.Simulate.blur(field);

        expect(component.props.onBlur).toBeCalled();
    });

    it("does not show the undo button if showUndo is not set to true", function () {
        var component = getComponent();

        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        ReactTestUtils.Simulate.change(field, { target: { value: "abc" } } );
        var undo = TestUtils.findRenderedDOMNodeWithDataId(component, "undo");
        expect(ReactTestUtils.isDOMComponent(undo)).toBeFalsy();
    });

    it("shows the undo button if showUndo is set to true", function () {
        var component = getComponent({ showUndo: true });

        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        ReactTestUtils.Simulate.change(field, { target: { value: "abc" } } );
        var undo = TestUtils.findRenderedDOMNodeWithDataId(component, "undo");
        expect(ReactTestUtils.isDOMComponent(undo)).toBeDefined();
    });

    it("shows the error message when it is specified", function () {
        var errorMessage = "help!";
        var component = getComponent({
            errorMessage: errorMessage
        });

        var errorDiv = TestUtils.findRenderedDOMNodeWithDataId(component, "form-text-area-errorMessage") ||
            TestUtils.findRenderedDOMNodeWithDataId(component, "form-text-area_errormessage");
        expect(errorDiv.textContent).toBe(errorMessage);
    });

    it("is disabled when it is specified", function () {
        var component = getComponent({ disabled: true });

        var textarea = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        expect(ReactTestUtils.isDOMComponent(textarea)).toBeTruthy();
        expect(textarea.disabled).toBeTruthy();
    });

    it("uses autoComplete when specified", function () {
        var component = getComponent({ useAutocomplete: true });

        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");

        expect(field.getAttribute("autocomplete")).toBe("on");
    });

    it("shows field as required", function () {
        var component = getComponent({ required: true });

        // verify that the component is rendered
        var field = TestUtils.findRenderedDOMNodeWithClass(component, "required");
        expect(ReactTestUtils.isDOMComponent(field)).toBeTruthy();
    });

    it("stateless: renders the component with default data-id", function () {
        var component = getComponent();

        var field = TestUtils.findRenderedDOMNodeWithDataId(component, "form-text-area");

        expect(field).toBeDefined();
    });

    it("stateless: renders the component with given data-id", function () {
        var component = getComponent({ "data-id": "myField" });

        var field = TestUtils.findRenderedDOMNodeWithDataId(component, "myField");

        expect(field).toBeDefined();
    });

    it("stateful: assigns empty string to value state when no value or defaultValue given", function () {
        const component = getComponent();
        const stateContainer = ReactTestUtils.findRenderedComponentWithType(component, StateContainer);

        expect(stateContainer.state.value).toBe("");
    });

    it("stateful: does not show the undo button if the originalValue param is not passed in", function () {
        var originalValue = "my original value";
        var component = getComponent({
            stateless: false,
            defaultValue: originalValue
        });

        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        ReactTestUtils.Simulate.change(field, { target: { value: "abc" } } );
        var undo = TestUtils.findRenderedDOMNodeWithDataId(component, "undo");
        expect(ReactTestUtils.isDOMComponent(undo)).toBeFalsy();
    });

    it("stateful: shows the undo icon when text changes", function () {
        const originalValue = "my original value";
        const component = getComponent({
            initialState: { value: originalValue },
            originalValue: originalValue,
            showUndo: true,
        });

        const field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        ReactTestUtils.Simulate.change(field, { target: { value: "abc" } } );
        const undo = TestUtils.findRenderedDOMNodeWithDataId(component, "undo");
        expect(ReactTestUtils.isDOMComponent(undo)).toBeTruthy();
    });

    it("stateful: reverts the input text to its original value if the undo icon is clicked", function () {
        var originalValue = "my original value";
        var component = getComponent({
            stateless: false,
            defaultValue: originalValue,
            originalValue: originalValue,
            showUndo: true,
        });

        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        ReactTestUtils.Simulate.change(field, { target: { value: "abc" } } );
        // check that the undo icon gets displayed
        var undo = TestUtils.findRenderedDOMNodeWithDataId(component, "undo");
        expect(ReactTestUtils.isDOMComponent(undo)).toBeTruthy();
        // click on the undo icon and verify that the field gets reverted to the original value
        ReactTestUtils.Simulate.click(undo);
        expect(field.value).toEqual(originalValue);
        // now check that the undo icon dissapeared
        undo = TestUtils.findRenderedDOMNodeWithDataId(component, "undo");
        expect(ReactTestUtils.isDOMComponent(undo)).toBeFalsy();
    });

    it("stateful: fires the onValueChange callback when clicking on the undo icon", function () {
        var originalValue = "my original value";
        var component = getComponent({
            stateless: false,
            defaultValue: originalValue,
            originalValue: originalValue,
            showUndo: true,
        });

        // make the undo icon appear by changing the field
        var field = TestUtils.findRenderedDOMNodeWithTag(component, "textarea");
        ReactTestUtils.Simulate.change(field, { target: { value: "abc" } } );
        // check that the icon is actually there
        var undo = TestUtils.findRenderedDOMNodeWithDataId(component, "undo");
        expect(ReactTestUtils.isDOMComponent(undo)).toBeTruthy();
        // click on the undo icon
        ReactTestUtils.Simulate.click(undo);
        // now we can verify that the callback gets triggered
        expect(component.props.onValueChange).toBeCalled();
    });

    /*
    TODO: fix this test
    NOTE: Currently we've not been able to test the react-tooltip helpHint in it's shown state. Since the React Tooltip
    does not populate its content until it is displayed, the content of the tooltip cannot yet be tested.
    */
    it("shows labelHelpText with the labelClassName applied", function () {
        var helpText = "help!";
        var component = getComponent({
            labelText: "label",
            labelHelpText: helpText,
            helpClassName: "bottom right",
            showUndo: true,
        });

        var help = TestUtils.findRenderedDOMNodeWithDataId(component, "help-tooltip");

        expect(help).toBeTruthy();
        // expect(help.getAttribute("class")).toContain("bottom right");
    });

    it("throws error when deprecated prop 'controlled' is passed in", function () {
        var expectedError = new Error(Utils.deprecatePropError("controlled", "stateless", "false", "true"));

        expect(function () {
            getComponent({ controlled: true });
        }).toThrow(expectedError);
    });

    it("fires cannonball warning when p-stateful flag isn't set", function () {
        console.warn = jest.fn();
        getComponent({ flags: [] });
        expect(console.warn).toBeCalled();
    });

    it("doesn't fire cannonball warning when p-stateful flag is set", function () {
        console.warn = jest.fn();
        getComponent({ flags: ["p-stateful"] });
        expect(console.warn).not.toBeCalled();
    });

    it("renders with no monospaced class when monospaced is false", () => {
        const component = mount(
            <FormTextArea
                monospaced={false}
            />
        );

        const spacing = component.find(".input-textarea__input--monospaced");

        expect(spacing.exists()).toEqual(false);
    });

    it("renders with single spacing class when spacing property is set to single", () => {
        const component = mount(
            <FormTextArea
                monospaced
            />
        );

        const spacing = component.find(".input-textarea__input--monospaced");

        expect(spacing.exists()).toEqual(true);
    });

    it("does not block resizing when noResize is false", () => {
        const component = mount(
            <FormTextArea
                noResize={false}
            />
        );
        const textarea = component.find(".input-textarea__input--no-resize");
        expect(textarea.exists()).toBeFalsy();
    });
    it("blocks resizing when noResize is true", () => {
        const component = mount(
            <FormTextArea
                noResize={true}
            />
        );
        const textarea = component.find("textarea.input-textarea__input--no-resize");
        expect(textarea.exists()).toBeTruthy();
    });
});
