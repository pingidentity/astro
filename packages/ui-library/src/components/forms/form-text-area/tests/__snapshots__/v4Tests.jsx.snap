// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTextArea v4 data-id's don't change 1`] = `
Array [
  "form-text-area",
  "form-text-area-textarea",
]
`;
